File: 2018/2/2_2.py
def commons(lines):
    size = len(lines)

    for i in range(size):
        line_1 = lines[i].strip()
        for j in range(i + 1, size):
            line_2 = lines[j].strip()

            diff = 0
            common = []
            for k in range(len(line_1)):
                if line_1[k] == line_2[k]:
                    common.append(line_1[k])
                else:
                    diff += 1

            if diff == 1:
                return common

    raise Exception('No common found')


with open('./input') as lines:
    print(''.join(commons(list(lines))))

File: 2018/2/2_2.clj
(use 'clojure.java.io)

(defn common-letters [id1 id2]
  (map first
       (filter #(= (nth % 0) (nth % 1))
               (map vector id1 id2))))

(defn close-enough?
  ([id1 id2  ] (close-enough? id1 id2 1))
  ([id1 id2 n] (<= (- (count id1)
                      (count (common-letters id1 id2)))
                   n)))

(defn couples [ids]
  (if
      (= (count ids ) 0)
    '()
    (let [h (first ids)
          l (rest ids)]
      (concat (map #(list h %) l)
              (couples l)))))

(defn close-commons [ids] 
  (filter #(apply close-enough? %) (couples ids)))

(with-open [r (reader "./input")]
  (->> r
       line-seq
       close-commons
       (map #(apply str %))
       println))
